### PR TITLE
endpoint: Set endpoint state to "ready" also after failure.

### PR DESCRIPTION
**Pull request by jrajahalme**
Set endpoint's state to "ready" also after failure, but make sure
endpoint's Status is not chnaged to "OK" in that case, so that cilium
API will return endpoint's state as "not-ready".

Move setting the endpoint's realized policy revision later to the
point when we know the build succeeded.

Fixes: #2069
Reported-by: Ian Vernon <ian@covalent.io>
Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

### CODE REVIEW

**Review by tgraf — APPROVED** [review]

File: pkg/endpoint/policy.go [hunk 1]
```diff
@@ -529,7 +529,18 @@ func (e *Endpoint) regenerate(owner Owner) error {
 		return fmt.Errorf("Failed to create endpoint directory: %s", err)
 	}
 
-	if err := e.regenerateBPF(owner, tmpDir); err != nil {
+	defer func() {
+		// Set to Ready, but only if no other changes are pending.
```
Comment (tgraf) [inline pkg/endpoint/policy.go]: I thought I understood the reason when I approved this but when Andre asked me today I could not recall. can you clarify this comment to explain why we need to go into ready state even on failure?

**Review by tgraf — CHANGES_REQUESTED** [review]

File: pkg/endpoint/policy.go (hunk 1 above)
Comment (jrajahalme) [reply pkg/endpoint/policy.go]: "ready" state is reported by the CLI as "not-ready" of the Status is not OK. Note that this PR also removes the wrong setting of Status to OK in the failure case.

Also, we do try to roll back to the previous policy, even though the bpf map changes are not currently rolled back.

current behavior is to remain in the regenerating state until a future change leads to successful regeneration. This PR does not change the end result in that case, as successful regeneration will also set the status to OK, allowing CLI to report the state as "ready".

Comment (jrajahalme) [reply pkg/endpoint/policy.go]: This was accompanied by a CLI feature that policy wait does wait forever. This I addressed in a separate PR.

**Comment by jrajahalme** [comment]
Rebase to master.

**Review by tgraf — APPROVED** [review]

**Pull request merged**